### PR TITLE
fix: validate Turnkey configuration at boundary

### PR DESCRIPTION
**Pull request by 0xgleb**
## What

Validate Turnkey organization IDs and API authentication keys at the configuration boundary instead of deferring malformed values to the signing SDK.

## Why

Whitespace-only organization IDs and malformed P-256 private keys could previously deserialize into apparently valid configuration and fail only when constructing or using the Turnkey client. That delayed startup failure, weakened error clarity, and allowed invalid states to exist in memory.

## How

- Parse `TurnkeyOrganizationId` through a fallible smart constructor.
- Trim surrounding organization-ID whitespace and reject an empty result.
- Parse `TurnkeyApiPrivateKey` through `TurnkeyP256ApiKey` and reject invalid P-256 key material.
- Keep API private-key `Debug` output redacted.
- Surface failures as configuration-parse errors before wallet construction.
- Gate signer-specific test imports so feature-minimal test builds remain valid.

## Testing

Added coverage for:

- trimming valid organization IDs;
- rejecting whitespace-only organization IDs;
- rejecting malformed P-256 API keys; and
- rejecting both invalid values through the full `WalletKind::Turnkey` configuration path.

GitHub Actions static checks and backend build/test matrix pass.

## Screenshots

Not applicable.

## Anything else

The deprecated infallible constructors remain temporarily for compatibility, while all configuration deserialization uses the validated path.

### CODE REVIEW

**Comment by 0xgleb** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.liquidity/1034?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#1029** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.liquidity/1029?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1037** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.liquidity/1037?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1034** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.liquidity/1034?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.liquidity/1034?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#1051** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.liquidity/1051?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1050** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.liquidity/1050?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1048** <a href="https://app.graphite.com/github/pr/ST0x-Technology/st0x.liquidity/1048?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`
<hr />

## How to use the Graphite Merge Queue

Add the label <kbd>add-to-gt-merge-queue</kbd> to this PR to add it to the merge queue.

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.com/invite/github/ST0x-Technology?ref=merge-queue-instructions-in-stack-comment&prId=1034).

<h2></h2>
<sub>

An organization admin has required the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: crates/evm/src/turnkey.rs [hunk 1]
```diff
@@ -42,33 +43,89 @@ use crate::{Evm, EvmError, TryIntoWallet, Wallet, WalletCtx};
...
+        value.try_into()
+    }
+
+    #[deprecated(note = "use TurnkeyApiPrivateKey::try_new")]
     pub fn new(value: String) -> Self {
         Self(value)
     }
 }
 
+impl TryFrom<String> for TurnkeyApiPrivateKey {
```
Comment (JuaniRios) [inline crates/evm/src/turnkey.rs:106]: critical: this validator panics instead of returning an error for the most likely bad input. `TurnkeyP256ApiKey::from_strings` hex-decodes the string and then calls `SecretKey::from_bytes(bytes.as_ref().into())`. That `.into()` is generic-array's `From<&[T]> for &GenericArray<T, N>`, which asserts the slice length is exactly 32 and panics otherwise. Only non-hex input takes the `Err` path. Reproduced against the locked versions: `from_strings("not-a-p256-private-key")` returns `Err`, but `from_strings("deadbeef")`, `"00"`, and `""` all panic with `assertion left == right failed`. The rustdoc on `try_new` promises an `InvalidApiPrivateKey` error. Because this now runs inside a serde `Deserialize` impl, a truncated or padded secret aborts the process at boot with a generic-array assertion that never mentions Turnkey, the wallet, or the config field. `validate-config` also cannot catch it, because the deploy gate keeps the wallet section as an untyped `toml::Value` and never deserializes these credentials, so the deploy passes and the host crash-loops. Please hex-decode yourself first, reject a length other than 32 with a dedicated error variant that carries the observed length and never the key material, and only then hand the bytes to the SDK. Apply the same guard at the `TracingTurnkeyClient::from_api_key` call site.

File: crates/evm/src/turnkey.rs [hunk 2]
```diff
@@ -730,6 +787,32 @@ mod tests {
...
+    }
+
+    #[test]
+    fn organization_id_rejects_only_whitespace() {
+        let error = TurnkeyOrganizationId::try_new("   ".to_string()).unwrap_err();
+
+        assert!(matches!(error, TurnkeyValueError::EmptyOrganizationId));
+    }
+
+    #[test]
```
Comment (JuaniRios) [inline crates/evm/src/turnkey.rs:1291]: should fix: both new negative tests feed `"not-a-p256-private-key"`, which is the only input class that returns an error. Every valid-hex wrong-length input panics instead, so the suite reports the validator as working while the dominant real failure mode aborts the process. The `wallet_kind_turnkey_*` tests also assert only `matches!(error, EvmError::WalletConfigParse(_))`, and an unrelated missing field produces the same variant, so they do not pin down why parsing failed. Please add cases for `"deadbeef"`, an empty string, and a 31-byte hex string, asserting the specific error variant. In the lib.rs tests, assert on the boxed error's message so the assertion distinguishes key validation from a generic serde error.

File: crates/evm/src/turnkey.rs [hunk 3]
```diff
@@ -42,33 +43,89 @@ use crate::{Evm, EvmError, TryIntoWallet, Wallet, WalletCtx};
...
+
+    #[deprecated(note = "use TurnkeyApiPrivateKey::try_new")]
     pub fn new(value: String) -> Self {
         Self(value)
     }
 }
 
+impl TryFrom<String> for TurnkeyApiPrivateKey {
+    type Error = TurnkeyValueError;
+
```
Comment (JuaniRios) [inline crates/evm/src/turnkey.rs:108]: should fix: `.map_err(|_| TurnkeyValueError::InvalidApiPrivateKey)?` throws away the underlying `StamperError`, which carries the specific reason (bad hex, wrong byte length, invalid scalar). `InvalidApiPrivateKey` is a unit variant with no `#[source]`, so every failure mode collapses into the same static message. AGENTS.md calls this out under Error Type Design: prefer `#[from]` and `?`, preserve chains with `#[source]`, and use bare `.map_err` only when `From` cannot adapt the source. Operators lose the concrete cause of a startup failure, which works against this PR's own goal of clearer config-boundary errors. Please make the variant carry the source, for example `InvalidApiPrivateKey(#[source] turnkey_api_key_stamper::StamperError)`, and map with `.map_err(TurnkeyValueError::InvalidApiPrivateKey)?`.

File: crates/evm/src/turnkey.rs [hunk 4]
```diff
@@ -42,33 +43,89 @@ use crate::{Evm, EvmError, TryIntoWallet, Wallet, WalletCtx};
...
+    ///
+    /// # Errors
+    ///
+    /// Returns [`TurnkeyValueError::EmptyOrganizationId`] when `value`
+    /// contains only whitespace.
+    pub fn try_new(value: String) -> Result<Self, TurnkeyValueError> {
+        value.try_into()
+    }
+
+    #[deprecated(note = "use TurnkeyOrganizationId::try_new")]
```
Comment (JuaniRios) [inline crates/evm/src/turnkey.rs]: should fix: `TurnkeyOrganizationId::new` and `TurnkeyApiPrivateKey::new` stay `pub` with zero callers. The diff migrated every call site to `try_new` and `test_organization_id()`, and `st0x-evm` is workspace-internal, so there is nothing to stay compatible with. The PR's stated motivation is that malformed values allowed invalid states to exist in memory; these two functions still allow exactly that, and `TurnkeyApiPrivateKey::new` is the path that keeps the panicking parse reachable. A deprecation warning is easy to allow away. Please delete both functions in this PR, so these types are constructible only through the validated path.

File: crates/evm/src/turnkey.rs [hunk 5]
```diff
@@ -42,33 +43,89 @@ use crate::{Evm, EvmError, TryIntoWallet, Wallet, WalletCtx};
...
     pub fn new(value: String) -> Self {
         Self(value)
     }
 }
 
+impl TryFrom<String> for TurnkeyOrganizationId {
+    type Error = TurnkeyValueError;
+
+    fn try_from(value: String) -> Result<Self, Self::Error> {
```
Comment (JuaniRios) [inline crates/evm/src/turnkey.rs:75]: should fix: `try_from` trims and rejects empty strings, but it still accepts non-UUID values such as `"org-test-123"`. Turnkey organization IDs are UUID-shaped in the production configs, and the generated Turnkey schema types them as UUIDs. So a typoed org ID still deserializes, passes startup, and fails only on the first signing request. That is the delayed failure this PR is meant to remove. Please parse the trimmed value as a UUID, store the canonical form, and add a `WalletKind::Turnkey` test that rejects a non-empty non-UUID id. Update the invalid-key test to use a valid UUID so it keeps testing the key path.

File: crates/evm/src/lib.rs [hunk 6]
```diff
@@ -1542,6 +1546,69 @@ mod tests {
         );
     }
 
+    #[cfg(feature = "turnkey")]
+    #[tokio::test]
```
Comment (JuaniRios) [inline crates/evm/src/lib.rs:1703]: minor: both new tests call `Anvil::new().spawn()` and build a live HTTP provider, but `try_into_wallet` fails during settings and credentials parsing, before the provider is touched. The Anvil process is pure overhead on two tests whose subject is JSON deserialization. This follows the pre-existing sibling tests, so it is not introduced here, but the two new cases are a good place to stop paying for it. Please factor a helper that builds `WalletCtx` for parse-only failures without spawning Anvil.

File: crates/evm/src/turnkey.rs [hunk 7]
```diff
@@ -4,9 +4,12 @@
 //! enclaves for low-latency signing (50-100ms). Like
 //! [`RawPrivateKeyWallet`](super::local::RawPrivateKeyWallet), it wraps
 //! the base provider with a [`WalletFiller`] -- the only difference is
-//! the signer: `TurnkeySigner` (remote signing via Turnkey API) instead
+//! the signer: `TracingTurnkeySigner` (remote signing via Turnkey API) instead
 //! of `PrivateKeySigner` (local key).
 
+use std::sync::Arc;
```
Comment (JuaniRios) [inline crates/evm/src/turnkey.rs]: nit: this patch puts a blank line between the std imports and the alloy imports. AGENTS.md puts std and external crates in one group with no empty lines inside it. rustfmt does not catch this. Please remove the blank line.

**Review by JuaniRios — COMMENTED** [review]
The direction is right, but the validator does the opposite of what it promises for the most likely bad input. A valid-hex key of the wrong length panics inside generic-array rather than returning an error, and because this runs in a serde Deserialize impl it aborts the process at boot. validate-config cannot catch it, since the deploy gate keeps the wallet section untyped, so a bad secret passes the gate and crash-loops the host. The new tests only feed non-hex input, which is the one class that does return an error.

**Review by JuaniRios — CHANGES_REQUESTED** [review]
Requesting changes. The validator does the opposite of what it promises for the most likely bad input:

- A valid-hex key of the wrong length panics inside generic-array's length assertion instead of returning `InvalidApiPrivateKey`. Only non-hex input takes the error path. Reproduced against the locked versions. Because this now runs inside a serde `Deserialize` impl, a truncated or padded secret aborts the process at boot with an assertion that never mentions Turnkey or the config field.
- `validate-config` cannot catch it either, because the deploy gate keeps the wallet section as an untyped `toml::Value`. The deploy passes and the host crash-loops.
- The new tests only feed non-hex input, which is the single class that does return an error, so the suite reports the validator as working.

Please also delete the deprecated infallible constructors (zero callers, workspace-internal crate) and preserve the parser error as a `#[source]`.

**Review by rouzwelt — APPROVED** [review]

File: crates/evm/src/turnkey.rs (hunk 1 above)
Comment (0xgleb) [reply crates/evm/src/turnkey.rs:106]: Fixed in the submitted head. The key is hex-decoded, its observed length is checked by converting it to [u8; 32] before the SDK sees it, and the stamper boundary repeats the check. Valid-hex, wrong-length inputs now return a typed error without exposing key material or reaching the panicking conversion.

File: crates/evm/src/turnkey.rs (hunk 2 above)
Comment (0xgleb) [reply crates/evm/src/turnkey.rs:1291]: Added empty, deadbeef, 31-byte, and 33-byte cases with exact InvalidApiPrivateKeyLength assertions. Wallet parse-path tests also assert the specific key-validation message, so unrelated serde failures cannot satisfy them.

File: crates/evm/src/turnkey.rs (hunk 3 above)
Comment (0xgleb) [reply crates/evm/src/turnkey.rs:108]: Fixed. Hex decoding, wrong length, and invalid P-256 scalar are distinct typed variants, while the hex, UUID, and Stamper errors are retained as sources.

File: crates/evm/src/turnkey.rs (hunk 4 above)
Comment (0xgleb) [reply crates/evm/src/turnkey.rs]: Fixed. The unchecked public constructors were removed; organization IDs and API private keys now enter through fallible, validated constructors.

File: crates/evm/src/turnkey.rs (hunk 5 above)
Comment (0xgleb) [reply crates/evm/src/turnkey.rs:75]: Fixed. Organization IDs are parsed as UUIDs, stored in canonical form, and covered by non-UUID and canonicalization tests. Invalid-key tests now use a valid organization UUID.

File: crates/evm/src/lib.rs (hunk 6 above)
Comment (0xgleb) [reply crates/evm/src/lib.rs:1703]: Fixed. Parse-only wallet tests now use a lightweight provider helper and no longer spawn Anvil.

File: crates/evm/src/turnkey.rs (hunk 7 above)
Comment (0xgleb) [reply crates/evm/src/turnkey.rs]: Fixed. The std and external imports form one uninterrupted group, followed by the internal-import group.
